src/Map.ts

```diff
@@ -0,0 +1,261 @@
+			});
+		}
+
+		get size(): number {
+			return this._map.size;
+		}
+
+		clear(): void {
+			return this._map.clear();
+		}
+
+		delete(key: K): boolean {
+			return this._map.delete(key);
+		}
+
+		entries(): any[] {
+			return this._map.entries();
+		}
+
+		forEach(callback: (key: K, value: V, instance: Map<K, V>) => any, context?: {}): void {
+			return this._map.forEach(callback, context);
+		}
+
+		get(key: K): V {
+			return this._map.get(key);
+		}
+
+		has(key: K): boolean {
+			return this._map.has(key);
+		}
+
+		keys(): K[] {
+			return this._map.keys();
+		}
+
+		set(key: K, value: V): Native.Map<K, V> {
+			this._map.set(key, value);
+			return this;
+		}
+
+		values(): V[] {
+			return this._map.values();
+		}
+	}
+}
+
+module Shim {
+	/**
+	 * An shimmed implementation of the ES2015 Map specification using arrays
+	 * instead of iterators. The iterator protocol is not implemented.
+	 */
+	export class Map<K, V> {
+		private _keys: K[] = [];
+		private _values: V[] = [];
+
+		/*
+		 * An alternative to Array.prototype.indexOf using Object.is
+		 * to check for equality. See http://mzl.la/1zuKO2V
+		 */
+		private _indexOfKey(keys: K[], key: K): number {
+			for (var i = 0; i < keys.length; i++) {
+				if (is(keys[i], key)) {
+					return i;
+				}
+			}
+			return -1;
+		}
+
+		constructor(iterable?: any) {
+			if (!(this instanceof Map)) {
+				throw new TypeError('Constructor Map requires "new"');
+			}
+
+			if (iterable) {
+				for (let [key, value] of iterable) {
+					this.set(key, value);
+				}
+			}
+		}
+
+		get size(): number {
+			return this._keys.length;
+		}
+
+		clear(): void {
+			this._keys.length = this._values.length = 0;
+		}
+
+		delete(key: K): boolean {
+			var index = this._indexOfKey(this._keys, key);
+			if (index < 0) {
+				return false;
+			}
+			this._keys.splice(index, 1);
+			this._values.splice(index, 1);
+			return true;
+		}
+
+		entries(): any[] {
+			var entries: any[] = [];
+			this._keys.forEach(function (key: K) {
+				entries.push([key, this.get(key)]);
+			}, this);
+			return entries;
+		}
+
+		forEach(callback: (key: K, value: V, mapInstance: Map<K, V>) => any, context?: {}) {
+			// don't use this.entries to avoid second forEach call
+			this._keys.forEach(function (key, index) {
+				callback.call(context, key, this.get(key), this);
+			}, this);
+		}
+
+		get(key: K): V {
+			var index = this._indexOfKey(this._keys, key);
+			return index < 0 ? undefined : this._values[index];
+		}
+
+		has(key: K): boolean {
+			return this._indexOfKey(this._keys, key) > -1;
+		}
+
+		keys(): K[] {
+			return this._keys;
+		}
+
+		set(key: K, value: V): Shim.Map<K, V> {
+			var index = this._indexOfKey(this._keys, key);
+			index = index < 0 ? this._keys.length : index;
+			this._keys[index] = key;
+			this._values[index] = value;
+			return this;
+		}
+
+		values(): V[] {
+			return this._values;
+		}
+	}
+}
+
+@hasClass('map', Native.Map, Shim.Map)
+export default class Map<K, V> {
```

Comment:
This will no longer be needed when we simplify to always returning a basic implementation regardless of native support.